src/lib.rs

```diff
@@ -319,27 +319,39 @@
     redundant_field_names,
 ))]
 #![deny(missing_docs)]
+#![cfg_attr(not(feature = "std"), no_std)]
 
 #[macro_use]
 extern crate serde;
+#[cfg(not(feature = "std"))]
+extern crate alloc;
 #[cfg(feature = "preserve_order")]
 extern crate indexmap;
 extern crate itoa;
 extern crate ryu;
 
 #[doc(inline)]
-pub use self::de::{from_reader, from_slice, from_str, Deserializer, StreamDeserializer};
+#[cfg(feature = "std")]
+pub use self::de::from_reader;
+#[doc(inline)]
+pub use self::de::{from_slice, from_str, Deserializer, StreamDeserializer};
 #[doc(inline)]
 pub use self::error::{Error, Result};
 #[doc(inline)]
-pub use self::ser::{
-    to_string, to_string_pretty, to_vec, to_vec_pretty, to_writer, to_writer_pretty, Serializer,
-};
+pub use self::ser::{to_string, to_string_pretty, to_vec, to_vec_pretty, Serializer};
+#[cfg(feature = "std")]
+pub use self::ser::{to_writer, to_writer_pretty};
 #[doc(inline)]
 pub use self::value::{from_value, to_value, Map, Number, Value};
 
+#[cfg(not(any(feature = "std", feature = "alloc")))]
+compile_error!("std and alloc are both disabled");
+#[cfg(all(feature = "std", feature = "alloc"))]
+compile_error!("std and alloc are both enabled");
```

Comment:
This needs to not be an error. Please find a way to support the case where the same dependency graph contains one crate that only needs serde_json with alloc and a different crate needs serde_json with std. That's part of what it means for features to be additive. I believe the serde crate currently does this correctly.